### ISSUE TITLE
[Feature Request] PrivateLink Support

### ISSUE THREAD

**Issue by rifelpet** [kubernetes-sigs/aws-load-balancer-controller#1859]
We have a need to create PrivateLink Endpoint Services with our NLBs. A Service resource would correspond to one NLB as well as one Endpoint Service configured to use that NLB.

I imagine the annotation design would be something like:

```yaml
service.beta.kubernetes.io/aws-load-balancer-endpoint-service-acceptance-required: 'true'
service.beta.kubernetes.io/aws-load-balancer-endpoint-service-allowed-principals: '["arn:aws:iam::123456789012:root"]'
service.beta.kubernetes.io/aws-load-balancer-endpoint-service-private-dns-name: 'foobar.example.com'
```

I don't have a good idea for how to conditionally add the DNS verification record... typically external-dns would handle that but given that both the name and value are computed, it doesn't align as well with the external-dns pattern of reading the name from a spec or annotation and the value from a resource's status or annotation.

A stretch goal would be to support creating Endpoints in a specified set of VPCs but given that (in our case) many of the VPCs would be in different AWS accounts and aws-load-balancer-controller doesn't support using multiple credentials for multiple accounts, I understand if we don't want to incorporate that complexity directly into aws-load-balancer-controller.

Would a PR for supporting Endpoint Services be accepted in the load balancer controller? Or are there any existing plans to add something similar?

**Comment by fandujar** [comment]
@rifelpet I need something similar. I got some issues when deleting a load balancer associated with a VPC Endpoint Service.

Maybe it would worth adding an annotation to your request to reuse a VPC Endpoint Service:
`service.beta.kubernetes.io/aws-load-balancer-endpoint-service-id: 'vpce-svc-xxxxxx'`

That way you can manage your VPC Endpoint Service without add complexity to aws-load-balancer-controller and still manage this association lifecycle of a load balancer with private link.

**Comment by M00nF1sh** [comment]
@fandujar @rifelpet 

We don't have plans internally for this. PRs are always welcome, however given this is a little complicated, i think we need a placeholder PR finalize the interfaces first before any code implementations, which we can review internally along with community review.

Personally i'd favor the `aws-load-balancer-endpoint-service-id` options if we can create a endpointService without NLB. (however seems an NLB is required when create a endpointService. I'm not familiar with privateLink so i might be wrong).

For the DNS validation, i don't think we can do much from the controller as it don't have control over the DomainNames. (we could modify it if it's owned by route53 like coreDNS, but seems it's over complicated for this controller's purpose). how about expose the validation records via annotation?

**Comment by rifelpet** [comment]
You are correct that `ec2.CreateVpcEndpointServiceConfiguration` requires that either an NLB or GWLB is provided, though you can later disassociate them from the endpoint service, resulting in an endpoint service with no load balancers behind it.

I think an `endpoint-service-id` annotation would only be useful for the following workflows:

1. Create Service resource in k8s (load-balancer-controller creates NLB)
2. Create VPC Endpoint Service externally, providing the new NLB's ARN
3. Add `endpoint-service-id` annotation to existing Service resource
4. When Service resource is deleted, aws-load-balancer-controller first detaches it from the endpoint service

or

1. Create VPC Endpoint Service externally attaching to some other NLB
2. Create Service resource in k8s with `endpoint-service-id` annotation for the new NLB to be attached to the existing endpoint service
3. When Service resource is deleted, aws-load-balancer-controller first detaches the NLB from the endpoint service

The first use case would likely be made easier if the aws-load-balancer-controller could manage the entire lifecycle of the endpoint service, which is what my original feature request addresses. The second use-case would be useful for endpoint services with many NLBs attached.

I'm going to plan on moving forward with my proposal for having aws-load-balancer-controller creating and managing the endpoint service resources, though much of the work could be reused in supporting an `endpoint-service-id` annotation. I'll have an interface PR draft open soon for review. I agree that aws-load-balancer-controller shouldn't handle the DNS validation but could expose that information via annotations.

**Comment by M00nF1sh** [comment]
did a second though, i agree with you. personally  i think we can expose the created service-id via annotation instead of validation informations. users can have automation tools to read the validation from the service-id

**Comment by rifelpet** [comment]
/remove-lifecycle stale

@M00nF1sh any update on this or reviewing the interface proposed in the linked PR?

**Comment by Cyanopus** [comment]
This is a very important business case. Adding a private link as a single annotation is a great way to expose your service B2B without network-level trust.

**Comment by nijave** [comment]
It'd also be useful if removing load balancers/deleting PrivateLinks on delete is possible. Right now, if you add a PrivateLink and try to delete the load balancer in Kubernetes, it gets stuck deleting the load balancer due to the associated PrivateLink

**Comment by tariq1890** [comment]
> You are correct that ec2.CreateVpcEndpointServiceConfiguration requires that either an NLB or GWLB is provided, though you can later disassociate them from the endpoint service, resulting in an endpoint service with no load balancers behind it.

@rifelpet @M00nF1sh  I think it's no longer required to have an NLB to create a VPC endpoint service. 

Source: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint_service#network-load-balancers

**Comment by rifelpet** [comment]
What in that link leads you to believe an NLB or GWLB are not required? the latest AWS CLI still requires it:

```
$ aws ec2 create-vpc-endpoint-service-configuration

An error occurred (InvalidParameter) when calling the CreateVpcEndpointServiceConfiguration operation: exactly one of network_load_balancer_arn or gateway_load_balancer_arn is a required member
```

**Comment by tariq1890** [comment]
I was going based off the terraform documentation. `acceptance_required` seems to be the only required field for the `aws_vpc_endpoint_service ` resource. Seems like the TF docs may be misleading

**Comment by rifelpet** [comment]
Yea it looks like each individual ARN field is optional but the docs don't mention that one of the two ARN fields must be provided

**Comment by valeriodelsarto** [comment]
Is the PR https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/2636 that aim to add this feature still WIP? Seems stale since a few months... having this feature will help a lot managing PrivateLink Endpoint Services from NLBs using only K8s resources

**Comment by eminugurkenar** [comment]
yes , having this is very useful for us too, would be good to have it on next releases.

**Comment by gosunilgo** [comment]
bumping this request -   What tasks are left to release-it / get-over-hump?   reminder pr #2636 from @valeriodelsarto

**Comment by kjenney** [comment]
This issue is blocking me at the moment. I see that the last comment in the PR was 2 days ago. Is there any way to get this PR merged ASAP?

**Comment by m17kea** [comment]
Any update on this?

**Comment by serhatcetinkaya** [comment]
any updates on this?

**Comment by igor-nikiforov** [comment]
This is really a missing feature. Managing PrivateLink externally on top of the NLBs created by the controller is such a pain.
